src/dev/mcp23017.h

```diff
@@ -0,0 +1,401 @@
+    INTCON_B
+    = 0x09, // Controls how the associated pin value is compared for the interrupt-on-change for port B.
+    IOCON  = 0x0A, // Controls the device.
+    GPPU_A = 0x0C, // Controls the pull-up resistors for the port A pins.
+    GPPU_B = 0x0D, // Controls the pull-up resistors for the port B pins.
+    INTF_A = 0x0E, // Reflects the interrupt condition on the port A pins.
+    INTF_B = 0x0F, // Reflects the interrupt condition on the port B pins.
+    INTCAP_A
+    = 0x10, // Captures the port A value at the time the interrupt occured.
+    INTCAP_B
+    = 0x11, // Captures the port B value at the time the interrupt occured.
+    GPIO_A = 0x12, // Reflects the value on the port A.
+    GPIO_B = 0x13, // Reflects the value on the port B.
+    OLAT_A = 0x14, // Provides access to the port A output latches.
+    OLAT_B = 0x15, // Provides access to the port B output latches.
+};
+
+enum class MCPPort : uint8_t
+{
+    A = 0,
+    B = 1
+};
+
+inline MCPRegister operator+(MCPRegister a, MCPPort b)
+{
+    return static_cast<MCPRegister>(static_cast<uint8_t>(a)
+                                    + static_cast<uint8_t>(b));
+}
+
+enum class MCPMode : uint8_t
+{
+    INPUT,
+    INPUT_PULLUP,
+    OUTPUT,
+};
+
+/**
+ * Barebones driver for MCP23017 I2C 16-Bit I/O Expander
+ * For now it supports only polling approach.
+ * 
+ * Usage:
+ *  Mcp23017 mcp;
+ *  mcp.Init();
+ *  mcp.PortMode(MCP23017Port::A, 0xFF); // Inputs
+ *  mcp.PortMode(MCP23017Port::B, 0xFF);
+ *  mcp.Read();
+ *  mcp.GetPin(2);
+ */
+class Mcp23017Transport
+{
+  public:
+    struct Config
+    {
+        I2CHandle::Config i2c_config;
+        uint8_t           i2c_address;
+        void              Defaults()
+        {
+            i2c_config.periph         = I2CHandle::Config::Peripheral::I2C_1;
+            i2c_config.speed          = I2CHandle::Config::Speed::I2C_1MHZ;
+            i2c_config.mode           = I2CHandle::Config::Mode::I2C_MASTER;
+            i2c_config.pin_config.scl = {DSY_GPIOB, 8};
+            i2c_config.pin_config.sda = {DSY_GPIOB, 9};
+            i2c_address               = 0x27;
+        }
+    };
+
+    void Init()
+    {
+        Config config;
+        config.Defaults();
+        Init(config);
+    };
+
+    void Init(const Config& config)
+    {
+        i2c_address_ = config.i2c_address << 1;
+        i2c_.Init(config.i2c_config);
+
+        //BANK = 	0 : sequential register addresses
+        //MIRROR = 	0 : use configureInterrupt
+        //SEQOP = 	1 : sequential operation disabled, address pointer does not increment
+        //DISSLW = 	0 : slew rate enabled
+        //HAEN = 	0 : hardware address pin is always enabled on 23017
+        //ODR = 	0 : open drain output
+        //INTPOL = 	0 : interrupt active low
+        WriteReg(MCPRegister::IOCON, 0b00100000);
+
+        //enable all pull up resistors (will be effective for input pins only)
+        WriteReg(MCPRegister::GPPU_A, 0xFF, 0xFF);
```

Comment:
This seems like it should be a part of the Device Initialization, and not a part of the Transport Initialization (since you'd probably want this to happen whether you were using the SPI or I2C versions of the device).